### ISSUE TITLE
Cannot query field "admin_notification" on type "Subscription"  when multiple projects defined in .graphqlrc.yml

### ISSUE THREAD

**Issue by waterdrop01** [dotansimha/graphql-eslint#526]
**Describe the bug**

Here is my `.graphqlrc.yml`:
```
projects:
  first:
    schema: ./first.graphql
    documents: ./src/first-gql.js
  second:
    schema: ./second.graphql
    documents: ./src/second-gql.js
```

Here is my `./src/first-gql.js` file:
```
const queryUsers = gql`
  query users($first: Int) {
    users(first: $first) {
      id
      symbol
    }
  }
`;
```

Here is my `./src/second-gql.js` file:
```
const subscribeAdminNotification = gql`
  subscription admin_notification(
    $limit: Int
    $order_by: [admin_notification_order_by!]
    $where: admin_notification_bool_exp
  ) {
    admin_notification(limit: $limit, order_by: $order_by, where: $where) {
      id
      level
    }
  }
`;
```
When running `npx eslint --ext graphql,js .` the `graphql-eslint` linter complains that `Cannot query field "admin_notification" on type "Subscription"`, because it is looking in the wrong `.graphql` schema!!!

If I remove the first schema definition from `.graphqlrc.yml`, then it works OK.

**Expected behavior**

`graphql-eslint` should lint .js files against their related .graphql schema definition (as defined by the `documents` field in `.graphqlrc.yml`)

Otherwise, it breaks compatibility with `graphql-config@3` which documentation states that [it's supporting multiple projects](https://graphql-config.com/usage#projects)

**May I know how do you lint when you have multiple .graphql files/projects in `.graphqlrc.yml`?**

Thanks for your help!

**Comment by B2o5T** [comment]
Hi @waterdrop01 thank you for the detailed reports!
@dotansimha I think currently it's not possible to handle multischemas projects?

**Comment by waterdrop01** [comment]
Here is the [`graphql-config@3` documentation that is referring to multiple projects support](https://graphql-config.com/usage#projects)

**Comment by dotansimha** [comment]
I think multi-project should work, since we are using `graphql-config`, and it does support. 
We are using `getProjectForFile` to get the correct schema per each file we are running `graphql-eslint` on, so if the file is included correctly in the project of `graphql-config`, it should load the schema correctly. 

I think there might be an issue here related to virtual files loaded from code-files? @B2o5T what do you think? I think that [here](https://github.com/dotansimha/graphql-eslint/blob/master/packages/plugin/src/schema.ts#L41) we get the virtual path, and then it leads to an an empty project, since the virtual file doesn't match the project? 

@waterdrop01 can you please try to put your GraphQL operation in a `.graphql` file and try to update your GraphQL-Config file accordingly? Just to verify that it only happens with virtual files.

**Comment by B2o5T** [comment]
Thank you @dotansimha ! I made a reproduction on my side and I can confirm that there was a problem due `virtual filepath`. I created PR to fix this :)

**Comment by waterdrop01** [comment]
I checked out your PR @B2o5T , but when doing `npm link` I get this bug => https://github.com/npm/arborist/issues/302

**Comment by waterdrop01** [comment]
i was in the root folder, seems like `npm link` should be run in the `packages/plugin` folder.

But still, in my graphql project, running the linter throws this error:

```
Oops! Something went wrong! :(

ESLint: 7.30.0

Error: ESLint configuration of processor in '.eslintrc.js#overrides[0]' is invalid: '@graphql-eslint/graphql' was not found.
    at ConfigValidator.validateProcessor (/home/me/code/github.com/project/repo/node_modules/@eslint/eslintrc/lib/shared/config-validator.js:233:19)
    at ConfigValidator.validateConfigArray (/home/me/code/github.com/project/repo/node_modules/@eslint/eslintrc/lib/shared/config-validator.js:321:18)
    at CascadingConfigArrayFactory._finalizeConfigArray (/home/me/code/github.com/project/repo/node_modules/@eslint/eslintrc/lib/cascading-config-array-factory.js:493:23)
    at CascadingConfigArrayFactory.getConfigArrayForFile (/home/me/code/github.com/project/repo/node_modules/@eslint/eslintrc/lib/cascading-config-array-factory.js:299:21)
    at FileEnumerator._iterateFilesWithFile (/home/me/code/github.com/project/repo/node_modules/eslint/lib/cli-engine/file-enumerator.js:365:43)
    at FileEnumerator._iterateFiles (/home/me/code/github.com/project/repo/node_modules/eslint/lib/cli-engine/file-enumerator.js:346:25)
    at FileEnumerator.iterateFiles (/home/me/code/github.com/project/repo/node_modules/eslint/lib/cli-engine/file-enumerator.js:296:59)
    at iterateFiles.next (<anonymous>)
    at CLIEngine.executeOnFiles (/home/me/code/github.com/project/repo/node_modules/eslint/lib/cli-engine/cli-engine.js:765:48)
    at ESLint.lintFiles (/home/me/code/github.com/project/repo/node_modules/eslint/lib/eslint/eslint.js:563:23)
```
What is the procedure to test `@graphql-eslint/eslint-plugin`?

**Comment by B2o5T** [comment]
@waterdrop01 you can try this:

`git checkout` this branch https://github.com/dotansimha/graphql-eslint/pull/532
`yarn`
`yarn build`
copy content of `packages/plugin/dist/index.cjs.js` to your `node_modules/@graphql-eslint/eslint-plugin/index.cjs.js`

And tell me if it solved your problem.

**Comment by dotansimha** [comment]
Or you can `npm pack` in the source directory and point Yarn/Npm to load the artifact `tgz`. 

@B2o5T if you'll push it and create the PR from a branch in this repo (instead of a fork), you'll get an alpha release from the PR.

**Comment by B2o5T** [comment]
@dotansimha thanks, I'll try for my next one 😆

**Comment by waterdrop01** [comment]
Ok, I confirm the fix works for me :+1: Awesome

**Comment by waterdrop01** [comment]
Well, actually directly linting the documents file works: `npx eslint ./query2.js`

BUT linting all files with `npx eslint --ext graphql,js .` still throw the error:

```
/home/me/code/github.com/waterdrop01/graphql-eslint-prettier/query2.js
  5:6  error  Cannot query field "book" on type "Query"  @graphql-eslint/fields-on-correct-type

✖ 1 problem (1 error, 0 warnings)
```

I created a [repository](https://github.com/waterdrop01/graphql-eslint-prettier) so you can reproduce the issue.

**Comment by waterdrop01** [comment]
This is probably due to the fact that in [`schema2.graphql`](https://github.com/waterdrop01/graphql-eslint-prettier/blob/master/schema2.graphql), the root Query name is redefined:

```
schema {
  query: query_root
}
```

**Comment by dotansimha** [comment]
> This is probably due to the fact that in [`schema2.graphql`](https://github.com/waterdrop01/graphql-eslint-prettier/blob/master/schema2.graphql?rgh-link-date=2021-07-15T11%3A29%3A54Z), the root Query name is redefined:
> 
> ```
> schema {
>   query: query_root
> }
> ```
> 
> So it's not `Query`, but `query_root`.

I don't we have a dependency directly for `"Query"`, we are using it based on the schema definition.

**Comment by waterdrop01** [comment]
Ok, I don't know then what could be the cause...What is reproducible using [this public repo](https://github.com/waterdrop01/graphql-eslint-prettier) is that:

- direct linting (`npx eslint ./query2.js`) of `js` files containing `gql` tags belonging to different graphql schema **works**
- global linting (`npx eslint --ext graphql,js .`) of `js` files containing `gql` tags belonging to different graphql schema **doesn't work**

This should explain why `vscode.graphql` extension doesn't work for me when `.graphqlrc` has multiple projects, as it's probably using a global expression (my guess).

**Comment by B2o5T** [comment]
Thank you so much @waterdrop01 for your reproduction repo, it is so much helpful to found where the issue comes!

The issue is comes from caching schemas. Currently, schema is cached by `filedir` of every file, and schema for `query2.js` file was incorrectly loaded due `query1.js` and `query2.js` are located in the same directory.
I changed caching schemas by real file path @dotansimha I think it's better.

**Comment by waterdrop01** [comment]
Wow, it just works! 

`graphql-eslint` is now correctly linting `gql` tags belonging to multiple projects.

And as a bonus, `vscode.graphql` extension now works perfectly.

Amazing, Thanks @B2o5T and @dotansimha for the fix and your precious help!

I let you close the issue when you feel appropriate :)

**Comment by waterdrop01** [comment]
Also, cross posting my _last_ remaining issue here as this may be somehow related: https://github.com/dotansimha/graphql-code-generator/issues/6282

If I set a remote graphql endpoint in `.graphqlrc.yml` like this:
```
projects:
  first:
    schema: https://first
    documents: ./src/first-gql.js
    extensions:
      codegen:
        generates:
        ./first.graphql:
          plugins:
            - schema-ast
      endpoints:
        default:
          url: https://first
  second:
    schema: https://second
    documents: ./src/second-gql.js
    extensions:
      codegen:
        generates:
        ./second.graphql:
          plugins:
            - schema-ast
      endpoints:
        default:
          url: https://second
```
 then code-generator `npx graphql-codegen generate --project first` works.

BUT then `graphql-eslint` doesn't work anymore (do not handle remote endpoint, only local schema files I guess...):
```
/home/me/code/github.com/project/repo/node_modules/node-fetch/lib/index.js:1222
                        throw new TypeError('Expected signal to be an instanceof AbortSignal');
                        ^

TypeError: Expected signal to be an instanceof AbortSignal
```
I may open a new issue for this.
But I'm not sure if it's `graphql-code-generator` faults or `graphql-eslint` that should be able to handle remote schema with multiple projects...

**Comment by B2o5T** [comment]
It seems like your issue is related to https://github.com/dotansimha/graphql-eslint/issues/525 that will be fixed soon

**Comment by dotansimha** [comment]
Fixed in `@graphql-eslint/eslint-plugin@1.1.4` :) Thanks @B2o5T ❤️
